File: docs/algorithms/code/algorithms/maxDepth.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
}

function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  const nodeQueue: Array<TreeNode> = [root];
  let ans = 0;

  while (nodeQueue.length) {
    let size = nodeQueue.length;

    while (size) {
      const node = nodeQueue.shift();

      if (node.left) {
        nodeQueue.push(node.left);
      }

      if (node.right) {
        nodeQueue.push(node.right);
      }

      size--;
    }

    ans += 1;
  }

  return ans;
}

interface NodeQueueItem {
  node: TreeNode;
  depth: number;
}

function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  const nodeQueue: Array<NodeQueueItem> = [
    {
      node: root,
      depth: 1,
    },
  ];

  let maxDepth = 1;

  while (nodeQueue.length) {
    const { node, depth } = nodeQueue.pop()!;

    if (node) {
      maxDepth = Math.max(maxDepth, depth);

      if (node.left) {
        nodeQueue.push({
          node: node.left,
          depth: depth + 1,
        });
      }

      if (node.right) {
        nodeQueue.push({
          node: node.right,
          depth: depth + 1,
        });
      }
    }
  }

  return maxDepth;
}

File: docs/algorithms/code/algorithms/maxDepth.go
// 思路：通过递归的方式，获取左右子树的最大深度，然后比较左右子树的最大深度，返回较大的深度
// 复杂度：时间复杂度 O(n) 需要遍历所有节点，空间复杂度 O(h) 递归栈的深度
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left := maxDepth(root.Left)
	right := maxDepth(root.Right)
	if left > right {
		return left + 1
	}
	return right + 1
}